pkg/nsx/services/securitypolicy/builder.go

```diff
@@ -598,22 +580,44 @@ func (service *SecurityPolicyService) buildRuleOutGroup(obj *v1alpha1.SecurityPo
 	return nsxRuleDstGroup, nsxRuleSrcGroupPath, nsxRuleDstGroupPath, nsxGroupShare, nil
 }
 
-func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int, createdFor string) string {
-	serializedBytes, _ := json.Marshal(rule)
-	ruleHash := fmt.Sprintf("%s", util.Sha1(string(serializedBytes)))
-	ruleIdxStr := fmt.Sprintf("%d", ruleIdx)
+func (service *SecurityPolicyService) buildRuleID(obj *v1alpha1.SecurityPolicy, ruleIdx int, createdFor string) string {
+	ruleIndexHash := service.buildRuleHashString(&(obj.Spec.Rules[ruleIdx]))
+
 	if IsVPCEnabled(service) {
-		suffix := strings.Join([]string{ruleIdxStr, ruleHash}, common.ConnectorUnderline)
-		return util.GenerateIDByObjectWithSuffix(obj, suffix)
+		ruleIndexHash = service.buildLimitedRuleHashString(&(obj.Spec.Rules[ruleIdx]))
+		return util.GenerateIDByObjectWithSuffix(obj, ruleIndexHash)
 	}
+
 	prefix := common.SecurityPolicyPrefix
 	if createdFor == common.ResourceTypeNetworkPolicy {
 		prefix = common.NetworkPolicyPrefix
 	}
-	return util.GenerateID(fmt.Sprintf("%s", obj.UID), prefix, ruleHash, ruleIdxStr)
+	ruleIdxSuffix := fmt.Sprintf("%d", ruleIdx)
+	return util.GenerateID(fmt.Sprintf("%s", obj.UID), prefix, ruleIdxSuffix, ruleIndexHash)
+}
+
+// A rule containing named port may be expanded to multiple NSX rules if the name ports map to multiple port numbers.
+// So, in VPC network, the rule port numbers, which either are defined in rule Port or resolved from named port, will be appended as CR rule baseID to distinguish them.
+// For T1, the portIdx and portAddressIdx are appended as suffix.
+func (service *SecurityPolicyService) buildExpandedRuleID(obj *v1alpha1.SecurityPolicy, rule *v1alpha1.SecurityPolicyRule, ruleIdx int,
```

Comment:
I saw you have remove parameter "rule" from function `buildRuleID` because we can use `obj *v1alpha1.SecurityPolicy` and `ruleIdx int` to calculate the rule inside the function, can we do it similarly in function `buildExpandedRuleID` to make it consistent?

Reply:
It makes sense to me to do this change. It's good to keep code consistent as uniformed style. 
Remove `rule *v1alpha1.SecurityPolicyRule` from `buildExpandedRuleID` function as well.